plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -265,6 +268,52 @@ long getLargeBlobThresholdInBytes() {
             {Storage.BlobWriteOption.doesNotExist(), Storage.BlobWriteOption.md5Match()};
     private static final Storage.BlobWriteOption[] OVERWRITE_CHECK_MD5 = {Storage.BlobWriteOption.md5Match()};
 
+    void writeBlob(String blobName, boolean failIfAlreadyExists, CheckedConsumer<OutputStream, IOException> writer) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, blobName).build();
+        final Storage.BlobWriteOption[] writeOptions = failIfAlreadyExists ? NO_OVERWRITE_NO_MD5 : OVERWRITE_NO_MD5;
+
+        StorageException storageException = null;
+
+        for (int retry = 0; retry < 3; ++retry) {
```

Comment:
I think the SDK already retries failed requests, is there a reason why we are retrying here too?

Reply:
Yea see the multi part upload logic. This has been an endless discussion back when it was introduced :) The short version of the story is this:

The chance of a multi-part/resumable write getting lost on GCPs end and returning a `410` is about 1/100k. Retrying would require buffering the full write beforehand which is not practical on the SDK layer => SDK will not retry this special case so we started doing it ourselves.